crates/blockchain-tree/src/blockchain_tree.rs

```diff
@@ -1237,14 +1234,8 @@ where
                     // State root calculation can take a while, and we're sure no write transaction
                     // will be open in parallel. See https://github.com/paradigmxyz/reth/issues/6168.
                     .disable_long_read_transaction_safety();
-                let (state_root, trie_updates) = StateRoot::from_tx(provider.tx_ref())
-                    .with_hashed_cursor_factory(HashedPostStateCursorFactory::new(
-                        DatabaseHashedCursorFactory::new(provider.tx_ref()),
-                        &hashed_state_sorted,
-                    ))
-                    .with_prefix_sets(prefix_sets)
-                    .root_with_updates()
-                    .map_err(Into::<BlockValidationError>::into)?;
+                let (state_root, trie_updates) = LatestStateProviderRef::new(&provider)
+                    .state_root_from_state_with_updates(hashed_state.clone())?;
```

Comment:
could this use `state_root_from_prefix_sets_with_updates`? I think this avoids the clone

Reply:
Unfortunately not as we need to overlay the updated state contained within `hased_state` on top of what is currently present in the database. `state_root_from_state_with_updates` reads state from the database and updates the trie paths defined in the provided `prefix_set: TriePrefixSets` argument. An alternative is to change the API and have it return `HashedPostStateSorted` as well. This would eliminate the need for the clone.

```rust
    fn incremental_state_root_with_updates(
        &self,
        range: std::ops::RangeInclusive<BlockNumber>,
    ) -> ProviderResult<(B256, TrieUpdates, HashedPostStateSorted)>
```

What do you think?

Reply:
Got it. That would work but don't want to create a huge diff with upstream.

Reply:
I've added a new method `state_root_from_state_with_updates_and_sorted_state` which removes the requirement for the clone. I think it's fine that we diverge slightly from upstream now. We can converge back to upstream when the rest of the `StateCommitment` PR's land.